### DISCUSSION TITLE
how is efficient encoding achieved with different cdf for different symbols

### DISCUSSION THREAD

**Discussion by qiutiantxwd** [InterDigitalInc/CompressAI#301] [Q&A] [answered]
Hello authors, 
Thanks for sharing this amazing code base!
This is a question on the general compress() function in EntropyModel Class. 
Take hyperprior model for example: 
Say the latent that needs to be encoded has dimension [C, H, W], then each of CxHxW symbols is assigned an separate estimated CDF. This CDF is either:
1. generated from a trainable network layer (like the hyper encoder output "z" in hyperprior model) ,
2. or indexed from a pre-generated table of CDFs (like the encoder output "y" in hyperprior model). 

My question is if each of all the CxHxW elements has their own estimated CDF, how does rANS achieve an efficient encoding? As I understand, ANS needs a single CDF (same for all incoming symbols) of a pre-defined alphabet to work. But here it looks like each symbol has their own CDF. 

I may have inaccurate understanding of ANS so would appreciate any pointers on this. 

Thanks!

**Comment by YodaEmbedding** [comment, marked as answer]
Arithmetic coding's primary benefit over other lossless encoding schemes is that you can change the encoding probability distribution on-the-fly at any time to almost anything you want. Just subdivide the interval differently.

ANS is an ["approximation of arithmetic coding"](https://en.wikipedia.org/wiki/Asymmetric_numeral_systems#Basic_concepts_of_ANS), and shares the same advantage of choosing new encoding distributions on-the-fly.
